File: hduoj/count_towers.cpp
#include <iostream>
using namespace std;

int main(){
	
	int d1[6][6],d2[6][6];
	int max,result;
	for (int i=1;i<=5;i++)
		for (int j=1;j<=i;j++)
		{
			cin>>d1[i][j];
			d2[i][j]=d1[i][j];
			
		}
		//构造数组d2
		int s=1;
		for(int x=5;x>=2;x--)
		{
			for (int y=1;y<=5-s;y++)
			{
				if(d2[x][y]>=d2[x][y+1])
					max=d2[x][y];
				else max=d2[x][y+1];
				
				//cout<<"max="<<max<<endl;
				d2[x-1][y]=max+d2[x-1][y];
			}
			s++;
		}

		//输出b2中的数据
		/*for (int i1=1;i1<=5;i1++)
		{
			for (int j1=1;j1<=i1;j1++)
			{
				cout<<d2[i1][j1]<<" ";	
			}
			cout<<endl;
		}*/
			
		//m横坐标，n纵坐标
		int m=1,n=1;
		
		cout<<"最大值的行走路径为："<<endl;
		//输出第一个路径
		cout<<d1[1][1]<<"  ";

		for(int a=1;a<=5;a++)
		{
			result=d2[m][n]-d1[m][n];
			if (result==d2[m+1][n])
			{
				m=m+1;//m的值每次都要向后移动一位
				cout<<d1[m][n]<<"  ";
			}
			else if(result==d2[m+1][n+1])
			{
				m=m+1;//m的值每次都要向后移动一位
				n=n+1;//n的值向后移动一位
				cout<<d1[m][n]<<"  ";
				
			}
		}
		cout<<endl;
        return 0;
}

File: hduoj/count_towers.java
import java.util.*;
public class count_towers {
    public static void main(String[] args) {
        Scanner cin = new Scanner(System.in);
        int [][] dp;
        int tests = cin.nextInt();
        for(int t = 0; t < tests; t++) {
            int lines = cin.nextInt();
            dp = new int[lines][lines];
            int [] arr = new int[100]; 
            for(int i = 0; i < lines; i++) {
               for(int j = 0; j <= i; j++) {
                   arr[j] = cin.nextInt();
               }
               if(i>0) dp[i][0] = dp[i - 1][0] + arr[0];//default int array value 0?
               else dp[i][0] = arr[0];
                   
               if(i > 1) 
                   dp[i][i] = dp[i - 1][i - 1] + arr[i];
               for(int j = 1; j < i; j++) {
                   dp[i][j] = (dp[i - 1][j - 1] > dp[i - 1][j]?dp[i - 1][j - 1]:dp[i - 1][j]) + arr[j];
               }
               

            }
            System.out.println(findMax(dp[lines - 1]));
        }
    }
    private static int findMax(int [] array) {
    int max = array[0];
    for (int i = 1; i < array.length; i++) {
        if (array[i] > max) {
          max = array[i];
        }
    }
    return max;
}
    private static void print2dArr(int [][] array) {
        System.out.println();
        for (int i = 0; i < array.length; i++) {
            for(int j = 0; j < array[0].length; j++) {
                System.out.print(array[i][j] + " ");
            }
            System.out.println();
        }
    }
}
